### ISSUE TITLE
Bad package:analyzer release? [0.37.1]

### ISSUE THREAD

**Issue by grouma** [dart-lang/sdk#37854]
Seeing these errors with dart --version:
Dart VM version: 2.5.0-dev.2.0 (Sun Aug 11 20:46:20 2019 +0200) on "linux_x64"

```
build_modules git:(more-prints) dart $(pub run build_runner generate-build-script) test --delete-conflicting-outputs -- -P presubmit
../../../.pub-cache/hosted/pub.dartlang.org/json_serializable-3.2.0/lib/src/field_helpers.dart:40:64: Error: The getter 'type' isn't defined for the class 'Element'.
 - 'Element' is from 'package:analyzer/dart/element/element.dart' ('../../../.pub-cache/hosted/pub.dartlang.org/analyzer-0.37.1/lib/dart/element/element.dart').
Try correcting the name to the name of an existing getter, or defining a getter or field named 'type'.
    final checkerA = TypeChecker.fromStatic(a.enclosingElement.type);
                                                               ^^^^
../../../.pub-cache/hosted/pub.dartlang.org/json_serializable-3.2.0/lib/src/field_helpers.dart:50:66: Error: The getter 'type' isn't defined for the class 'Element'.
 - 'Element' is from 'package:analyzer/dart/element/element.dart' ('../../../.pub-cache/hosted/pub.dartlang.org/analyzer-0.37.1/lib/dart/element/element.dart').
Try correcting the name to the name of an existing getter, or defining a getter or field named 'type'.
      final checkerB = TypeChecker.fromStatic(b.enclosingElement.type);
```

**Comment by grouma** [comment]
Looking into this it looks like `enclosingElement` now returns type `Element` instead of `ClassElement` and therefore `type` is not necessarily defined.

**Comment by devoncarew** [comment]
@stereotype441

**Comment by stereotype441** [comment]
I'm investigating

**Comment by stereotype441** [comment]
> Looking into this it looks like `enclosingElement` now returns type `Element` instead of `ClassElement` and therefore `type` is not necessarily defined.

Agreed.  It looks like in https://dart-review.googlesource.com/c/sdk/+/108504 we changed the return type of `ClassMemberElement.enclosingElement` to `Element` (presumably because with extension methods, elements derived from `ClassMemberElement` no longer necessarily represent class members, but could now represent members of extensions as well).

I can see a few possible fixes, and none of them is great:
- We could back out the breaking change (and any later changes that depend on it) and find a way to do it that's non-breaking.  But the change was a month ago, and the majority of the work that's been done on extension methods since then probably depends on it.  So it seems like this option is a non-starter.
- We could land a CL that restores the old API without breaking the extension methods functionality.  That probably looks something like:
  - Change the return type of `ClassMemberElement.enclosingElement` back to `ClassElement`.
  - Introduce a new getter `ClassMemberElement.enclosingElement2` with the return type of `Element`.
  - Change any call sites in the analyzer that need to work with extension methods so that they use `enclosingElement2`.
  - Note that `ClassMemberElement.enclosingElement` overrides `Element.enclosingElement`, so we might need to add `Element.enclosingElement2` as well.  And `Element.enclosingElement` is pretty widely used, so we might have a lot of call sites to switch over.
  - Later, in analyzer version 0.38.0, we would restore the type of `ClassMemberElement.enclosingElement` to `Element` (matching the behavior of the new `.enclosingElement2` getter) and deprecate `.enclosingElement2`.
- We could make a change that restores the old API and breaks extension methods, and publish it as anaylzer 0.37.1+1.  We couldn't land this change on master in the SDK, but we could do it in a branch.  Then we could hurry up and publish 0.38.0, and tell people, "if you want extension method support, you'll need to upgrade to analyzer 0.38.0".
- We could leave the analyzer as is, and force clients to publish their own "bug fixes" to adapt to our inadvertent breakage (kind of like what @grouma experimented with in https://github.com/dart-lang/json_serializable/pull/528).

I'll talk with the analyzer team tomorrow and we'll choose one of these possibilities, or come up with something better.

(Side note: it seems like unfortunate nomenclature if `ClassMemberElement.enclosingElement` isn't guaranteed to return a `ClassElement`; after all, if a thing isn't guaranteed to be inside a class, we probably shouldn't be calling it a "class member".  So we should probably also discuss whether we should do some API renaming as part of 0.38.0 to avoid user confusion)

**Comment by kevin-haynie** [comment]
+1.  Having the same problem.  This is breaking my code.

**Comment by devoncarew** [comment]
@kevin-haynie, is this for tthe same package (json_serializable) or a different one?

**Comment by kevin-haynie** [comment]
Yes - json_serializable.  I'm downgrading now to 1.8.3.

**Comment by kevin-haynie** [comment]
hmmm.  I just downgraded to 1.8.3 but I'm still having the same problem: the getter 'type' isn't defined for the class 'Element'.  Do I need to downgrade/restore something else to get back to a working state?

**Comment by grouma** [comment]
You can restrict your analyzer version to `<0.37.1`.

**Comment by jakemac53** [comment]
It sounds to me like we should be republishing 0.37.0 as 0.37.1+1 and then re-release as 0.38.0?

Yes that means the world has to roll forward before people can use extension methods, but I don't think that is unreasonable/unexpected.

In an ideal world the analyzer (or anything that parses dart code) would not use <3.0.0 sdk constraints but instead restrict to < the next feature release version. We can't change that now though. After Dart 3.0 we should really try to do this so we can actually version properly.

**Comment by stereotype441** [comment]
After discussion we decided to do this option:

> * We could make a change that restores the old API and breaks extension methods, and publish it as anaylzer 0.37.1+1.  We couldn't land this change on master in the SDK, but we could do it in a branch.  Then we could hurry up and publish 0.38.0, and tell people, "if you want extension method support, you'll need to upgrade to analyzer 0.38.0".

(@jakemac53, this is similar to your suggestion, but it has two advantages: (1) it allows 0.37.1+1 to retain the other improvements in 0.37.1, and (2) it allows us to roll out the other breaking changes we've had in our staging CL as part of 0.38.0, so that we minimize impact on the ecosystem by avoiding an unnecessary extra major version number bump)

I will publish 0.37.1+1 ASAP.  In the meantime @bwilkerson is going to make a CL that documents the fact that `ClassMemberElement` objects no longer are guaranteed to be inside a `ClassElement`, and we'll roll that into 0.38.0, which we'll publish as soon as possible (probably tomorrow).  Renaming `ClassMemberElement` will take some more doing, so we'll do it later, probably in analyzer version 0.39.0.

**Comment by stereotype441** [comment]
Candidate change for version 0.37.1+1 is here: https://dart-review.googlesource.com/c/sdk/+/113100

**Comment by stereotype441** [comment]
Ok, I've just published analyzer version 0.37.1+1, which should fix the problem.  Please feel to reopen if you experience any further issues.

**Comment by jakemac53** [comment]
Thanks!
